From 3e8594fc8437f84cd584956316199abd92cfc31a Mon Sep 17 00:00:00 2001
From: Sergey Dyasli <sergey.dyasli@citrix.com>
Date: Tue, 23 Jul 2019 16:49:29 +0100
Subject: [PATCH] xsm: hide detailed Xen version from unprivileged guests

These subops leak information into guests which they have no buisness
knowning.  Return empty strings instead.

Signed-off-by: Sergey Dyasli <sergey.dyasli@citrix.com>
diff --git a/xen/common/version.c b/xen/common/version.c
index 937eb1281c70..cc621ab76af4 100644
--- a/xen/common/version.c
+++ b/xen/common/version.c
@@ -67,7 +67,7 @@ const char *xen_banner(void)
 
 const char *xen_deny(void)
 {
-    return "<denied>";
+    return "";
 }
 
 static const void *build_id_p __read_mostly;
diff --git a/xen/include/xsm/dummy.h b/xen/include/xsm/dummy.h
index af96299340f2..864b899c4974 100644
--- a/xen/include/xsm/dummy.h
+++ b/xen/include/xsm/dummy.h
@@ -796,14 +796,22 @@ static XSM_INLINE int xsm_xen_version(XSM_DEFAULT_ARG uint32_t op)
     case XENVER_get_features:
         /* These sub-ops ignore the permission checks and return data. */
         return 0;
-    case XENVER_extraversion:
-    case XENVER_compile_info:
-    case XENVER_capabilities:
-    case XENVER_changeset:
+
     case XENVER_pagesize:
     case XENVER_guest_handle:
         /* These MUST always be accessible to any guest by default. */
         return xsm_default_action(XSM_HOOK, current->domain, NULL);
+
+    case XENVER_extraversion:
+    case XENVER_compile_info:
+    case XENVER_capabilities:
+    case XENVER_changeset:
+    case XENVER_commandline:
+        if ( IS_ENABLED(CONFIG_DEBUG) )
+            /* Expose information to guests only in debug builds. */
+            return xsm_default_action(XSM_HOOK, current->domain, NULL);
+
+        /* fallthrough */
     default:
         return xsm_default_action(XSM_PRIV, current->domain, NULL);
     }
